File: algorithms/bellman-ford.java
// The Bellman-Ford algorithm works by iteratively relaxing edges in the graph.
// It starts by assuming an infinite distance for all vertices except the source vertex, which has a distance of 0.
// Then, it relaxes each edge repeatedly, updating the shortest known distance to each vertex until no more updates can be made.

// Imagine you are managing a transportation network where cities are connected by roads with different travel times.
// The Bellman-Ford algorithm can be used to determine the shortest travel times from a starting city to all other cities, considering possible delays or negative weights due to traffic conditions.

import java.util.*;

class Edge {
    int src, dest, weight;
    Edge(int src, int dest, int weight) {
        this.src = src;
        this.dest = dest;
        this.weight = weight;
    }
}

public class BellmanFordAlgorithm {
    public static void bellmanFord(List<Edge> edges, int V, int source) {
        int[] dist = new int[V];
        Arrays.fill(dist, Integer.MAX_VALUE);
        dist[source] = 0;

        // Relax edges |V| - 1 times
        for (int i = 0; i < V - 1; i++) {
            for (Edge edge : edges) {
                if (dist[edge.src] != Integer.MAX_VALUE && dist[edge.src] + edge.weight < dist[edge.dest]) {
                    dist[edge.dest] = dist[edge.src] + edge.weight;
                }
            }
        }

        for (Edge edge : edges) {
            if (dist[edge.src] != Integer.MAX_VALUE && dist[edge.src] + edge.weight < dist[edge.dest]) {
                System.out.println("Graph contains negative weight cycle!");
                return;
            }
        }

        for (int i = 0; i < V; i++) {
            System.out.println("Shortest distance from source to vertex " + i + ": " + dist[i]);
        }
    }
    public static void main(String[] args) {
        int V = 5; // Number of vertices
        int source = 0; // Source vertex
        List<Edge> edges = new ArrayList<>();
        edges.add(new Edge(0, 1, 6));
        edges.add(new Edge(0, 2, 7));
        edges.add(new Edge(1, 2, 8));
        edges.add(new Edge(1, 3, -4));
        edges.add(new Edge(1, 4, 5));
        edges.add(new Edge(2, 3, 9));
        edges.add(new Edge(2, 4, -3));
        edges.add(new Edge(3, 1, 7));
        edges.add(new Edge(4, 0, 2));
        edges.add(new Edge(4, 3, 7));

        bellmanFord(edges, V, source);
    }
}

File: algorithms/bellman-ford.kt
// The Bellman-Ford algorithm is used to find the shortest path from a single source vertex to all other vertices in a weighted graph.
// It can handle negative weights, unlike Dijkstra's algorithm.
// The algorithm works by repeatedly relaxing the edges of the graph and minimizing the distance to each vertex from the start vertex over all possible paths.

// Imagine you are planning a road trip through several cities, and you're looking at a map with the time it takes to travel between each pair of cities.
// Some roads might have heavy traffic or be under construction, making travel slower.
// You want to find the shortest travel time from your starting city to each other city, considering that some paths might take longer than others.
// The Bellman-Ford algorithm is perfect for this scenario, especially since it can handle routes with negative travel times (e.g. faster shortcuts).

class Edge(val source: Int, val destination: Int, val weight: Int)

class Graph(private val vertices: Int, private val edges: List<Edge>) {
    fun bellmanFord(startVertex: Int): IntArray {
        val distance = IntArray(vertices) { Int.MAX_VALUE }

        for (i in 0 until vertices - 1) {
            for (edge in edges) {
                if (distance[edge.source] + edge.weight < distance[edge.destination]) {
                    distance[edge.destination] = distance[edge.soruce] + edge.weight
                }
            }
        }

        // Check for negative-weight cycles
        for (edge in edges) {
            if (distance[edge.source] + edge.weight < distance[edge.destination]) {
                println("Graph contains a negative-weight cycle")
                return distance
            }
        }

        return distance
    }
}
